File: problem3/solution.cpp
#include <iostream>
using namespace std;

struct ListNode{
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {};
    ListNode(int value) : val(value), next(nullptr) {};
    ListNode(int value, ListNode* nextNode) : val(value), next(nextNode) {};
};

class Solution{
    public:
        ListNode* mergeTwoLists(ListNode* list1, ListNode* list2){
            ListNode* dummy = new ListNode;
            ListNode* tail = dummy;

            while (list1 != nullptr && list2 != nullptr){
                if (list1->val < list2->val){
                    tail->next = list1;
                    list1 = list1->next;
                }else{
                    tail->next = list2;
                    list2 = list2->next;
                }
                tail = tail->next;
            }

            if (list1 != nullptr){
                tail->next = list1;
            }else{
                tail->next = list2;
            }
            ListNode* result = dummy->next;
            delete dummy;
            return result;
        }

};

int main(){
    ListNode* list1 = new ListNode(3, new ListNode(7, new ListNode(9)));
    ListNode* list2 = new ListNode(5, new ListNode(8, new ListNode(11)));
    Solution sol;

    ListNode* result = sol.mergeTwoLists(list1, list2);
    while (result != nullptr){
        cout << result->val << endl;
        result = result->next;
    }
    return 0;
}

File: problem3/Solution.java
class Solution {

    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode dummy = new ListNode();
        ListNode nextNode = dummy;
        while (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                nextNode.next = list1;
                list1 = list1.next; // Move list1 pointer forward
            } else {
                nextNode.next = list2;
                list2 = list2.next; // Move list2 pointer forward
            }
            nextNode = nextNode.next;
        }

        if (list1 != null) {
            nextNode.next = list1;
        } else if (list2 != null) {
            nextNode.next = list2;
        }
        return dummy.next;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        ListNode list1 = new ListNode(1, new ListNode(3, new ListNode(5)));
        ListNode list2 = new ListNode(2, new ListNode(4, new ListNode(6)));

        ListNode answer = solution.mergeTwoLists(list1, list2);
        while (answer != null) {
            System.out.println(answer.val);
            answer = answer.next;
        }
    }
}
